java.lang.ClassCastException: class org.apache.logging.slf4j.SLF4JLoggerContext cannot be cast to class org.apache.logging.log4j.core.LoggerContext (org.apache.logging.slf4j.SLF4JLoggerContext and org.apache.logging.log4j.core.LoggerContext are in unnamed module of loader 'app')

Error message from worker: java.lang.RuntimeException: org.apache.beam.vendor.guava.v26_0_jre.com.google.common.util.concurrent.UncheckedExecutionException: java.lang.ClassCastException: class org.apache.logging.slf4j.SLF4JLoggerContext cannot be cast to class org.apache.logging.log4j.core.LoggerContext (org.apache.logging.slf4j.SLF4JLoggerContext and org.apache.logging.log4j.core.LoggerContext are in unnamed module of loader 'app') org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory$1.typedApply(IntrinsicMapTaskExecutorFactory.java:197) org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory$1.typedApply(IntrinsicMapTaskExecutorFactory.java:168) org.apache.beam.runners.dataflow.worker.graph.Networks$TypeSafeNodeFunction.apply(Networks.java:67) org.apache.beam.runners.dataflow.worker.graph.Networks$TypeSafeNodeFunction.apply(Networks.java:54) org.apache.beam.runners.dataflow.worker.graph.Networks.replaceDirectedNetworkNodes(Networks.java:91) org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory.create(IntrinsicMapTaskExecutorFactory.java:128) org.apache.beam.runners.dataflow.worker.BatchDataflowWorker.doWork(BatchDataflowWorker.java:361) org.apache.beam.runners.dataflow.worker.BatchDataflowWorker.getAndPerformWork(BatchDataflowWorker.java:314) org.apache.beam.runners.dataflow.worker.DataflowBatchWorkerHarness$WorkerThread.doWork(DataflowBatchWorkerHarness.java:140) org.apache.beam.runners.dataflow.worker.DataflowBatchWorkerHarness$WorkerThread.call(DataflowBatchWorkerHarness.java:120) org.apache.beam.runners.dataflow.worker.DataflowBatchWorkerHarness$WorkerThread.call(DataflowBatchWorkerHarness.java:107) java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) java.base/java.lang.Thread.run(Thread.java:834) Caused by: org.apache.beam.vendor.guava.v26_0_jre.com.google.common.util.concurrent.UncheckedExecutionException: java.lang.ClassCastException: class org.apache.logging.slf4j.SLF4JLoggerContext cannot be cast to class org.apache.logging.log4j.core.LoggerContext (org.apache.logging.slf4j.SLF4JLoggerContext and org.apache.logging.log4j.core.LoggerContext are in unnamed module of loader 'app') org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2050) org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LocalCache.get(LocalCache.java:3952) org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4871) org.apache.beam.runners.dataflow.worker.UserParDoFnFactory.create(UserParDoFnFactory.java:95) org.apache.beam.runners.dataflow.worker.DefaultParDoFnFactory.create(DefaultParDoFnFactory.java:75) org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory.createParDoOperation(IntrinsicMapTaskExecutorFactory.java:267) org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory.access$000(IntrinsicMapTaskExecutorFactory.java:89) org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory$1.typedApply(IntrinsicMapTaskExecutorFactory.java:186) ... 14 more Caused by: java.lang.ClassCastException: class org.apache.logging.slf4j.SLF4JLoggerContext cannot be cast to class org.apache.logging.log4j.core.LoggerContext (org.apache.logging.slf4j.SLF4JLoggerContext and org.apache.logging.log4j.core.LoggerContext are in unnamed module of loader 'app') org.apache.logging.log4j.core.LoggerContext.getContext(LoggerContext.java:193) org.apache.logging.log4j.core.Logger$LoggerProxy.readResolve(Logger.java:665) java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) java.base/java.lang.reflect.Method.invoke(Method.java:566) java.base/java.io.ObjectStreamClass.invokeReadResolve(ObjectStreamClass.java:1265) java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2205) java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679) java.base/java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2464) java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2358) java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2196) java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679) java.base/java.io.ObjectInputStream.readArray(ObjectInputStream.java:2102) java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1667) java.base/java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2464) java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2358) java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2196) java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679) java.base/java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2464) java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2358) java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2196) java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679) java.base/java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2464) java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2358) java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2196) java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679) java.base/java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2464) java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2358) java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2196) java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679) java.base/java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2464) java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2358) java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2196) java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679) java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:493) java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:451) org.apache.beam.sdk.util.SerializableUtils.deserializeFromByteArray(SerializableUtils.java:75) org.apache.beam.runners.dataflow.worker.UserParDoFnFactory$UserDoFnExtractor.getDoFnInfo(UserParDoFnFactory.java:66) org.apache.beam.runners.dataflow.worker.UserParDoFnFactory.lambda$create$0(UserParDoFnFactory.java:98) org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4876) org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3528) org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2277) org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154) org.apache.beam.vendor.guava.v26_0_jre.com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044) ... 21 more

--- a/TUTORIAL.md
+++ b/TUTORIAL.md
@@ -3,6 +3,7 @@ Tips for running app.
 * TLS
   * [Creating self-signed certificate for elasticsearch](#creating-self-signed-certificate-for-elasticsearch)
   * [Extracting CA crt and key from ECK secrets](#extracting-ca-cert-and-key-from-eck-secrets)
+  * [Indexing documents to eck using dataflow runner](#indexing-documents-to-eck-using-dataflow-runner)
 
 ## Creating self-signed certificate for elasticsearch 
 
@@ -79,15 +80,48 @@ This section will extract CA key and cert and combine them as PKCS#12 keystore.
 
 1. Extract CA Cert from secret
 ```shell
-kubectl get secrets es-es-http-ca-internal -n YOUR_NAME_SPACE -o jsonpath='{.data.tls\.crt}' | base64 -d > ca.crt
+export KUBE_NS=elastic
+mkdir certs ; cd certs
+kubectl get secrets es-es-http-ca-internal -n $KUBE_NS -o jsonpath='{.data.tls\.crt}' | base64 -d > ca.crt
 ```
 
 2. Extract CA Key
 ```shell
-kubectl get secrets es-es-http-ca-internal -n YOUR_NAME_SPACE -o jsonpath='{.data.tls\.key}' | base64 -d > ca.key
+kubectl get secrets es-es-http-ca-internal -n $KUBE_NS -o jsonpath='{.data.tls\.key}' | base64 -d > ca.key
 ```
 
 3. Combine crt and key as PKCS#12 keystore.
 ```shell
 openssl pkcs12 -export -in ca.crt -inkey ca.key -out ca.p12 -name "Elastic CA certificate"
 ```
+
+## Indexing documents to ECK using dataflow runner
+
+This section will outline how to index documents to TLS protected elasticsearch running on ECK.  
+You will need to stage keystore including CA key and certificate inside dataflow _workers_.
+
+There are several options to stage TLS cert on _workers_.
+
+- Use [fileToStage option](https://cloud.google.com/dataflow/docs/reference/pipeline-options) and expose as _files_ on the worker.
+- Bundle it inside [custom container](https://cloud.google.com/dataflow/docs/guides/using-custom-containers)  and use it as runtime environment for user code
+
+I will use _custom container_ for simplicity.
+
+1. Prepare keystore following instruction in [previous section](#extracting-ca-cert-and-key-from-eck-secrets).
+
+2. Build custom container and push to container registry. See [Dockerfile](java/Dockerfile) for example.
+
+3. Run pipeline. `--experiments=use_runner_v2` and `--sdkContainerImage` option will be required.
+
+```shell
+e.g
+java -cp build/libs/java-1.0-SNAPSHOT-all.jar app.examples.ElasticsearchIOSimpleWrite\
+      --runner=DataflowRunner\
+      --gcpTempLocation=MY_GS_STAGING_LOCATION\
+      --project=MY_PROJECT\
+      --region=MY_REGION\
+      --experiments=use_runner_v2\
+      --sdkContainerImage=MY_PUSHED_IMAGE
+```
+
+4. Verify documents are indexed to elasticsearch

--- a/java/Dockerfile
+++ b/java/Dockerfile
@@ -1,0 +1,6 @@
+# Use image built by apache beam project
+FROM apache/beam_java11_sdk:2.42.0
+
+# Copy keystore to the image
+ARG KEYSTORE_NAME=ca.p12
+COPY certs/$KEYSTORE_NAME /tmp/$KEYSTORE_NAME

--- a/java/build.gradle
+++ b/java/build.gradle
@@ -15,6 +15,10 @@ ext {
 
 repositories {
     mavenCentral()
+
+    maven {
+        url = uri("https://packages.confluent.io/maven/")
+    }
 }
 
 dependencies {
@@ -23,6 +27,7 @@ dependencies {
     implementation(
             "org.apache.beam:beam-sdks-java-core:${beamVersion}",
             "org.apache.beam:beam-runners-direct-java:${beamVersion}",
+            "org.apache.beam:beam-runners-google-cloud-dataflow-java:${beamVersion}",
             "org.apache.beam:beam-sdks-java-io-elasticsearch:${beamVersion}",
             "org.apache.beam:beam-sdks-java-io-mongodb:${beamVersion}",
             "org.apache.logging.log4j:log4j-api:${log4jVersion}",
@@ -30,6 +35,7 @@ dependencies {
             "org.apache.logging.log4j:log4j-slf4j-impl:${log4jVersion}",
             "org.mongodb:mongodb-driver-sync:4.6.0"
     )
+
     annotationProcessor "com.google.auto.value:auto-value:${autoValueVersion}"
 }
 
@@ -54,4 +60,9 @@ test {
             SOME_KEY: "some_key"
     ]
     environment(envs)
+}
+
+shadowJar {
+    zip64 true
+    mergeServiceFiles()
 }

--- a/java/build.gradle
+++ b/java/build.gradle
@@ -9,7 +9,7 @@ version '1.0-SNAPSHOT'
 
 ext {
     beamVersion = "2.42.+"
-    log4jVersion = "2.19.+"
+    slf4jVersion = "2.0.+"
     autoValueVersion = "1.10.+"
 }
 
@@ -30,9 +30,8 @@ dependencies {
             "org.apache.beam:beam-runners-google-cloud-dataflow-java:${beamVersion}",
             "org.apache.beam:beam-sdks-java-io-elasticsearch:${beamVersion}",
             "org.apache.beam:beam-sdks-java-io-mongodb:${beamVersion}",
-            "org.apache.logging.log4j:log4j-api:${log4jVersion}",
-            "org.apache.logging.log4j:log4j-core:${log4jVersion}",
-            "org.apache.logging.log4j:log4j-slf4j-impl:${log4jVersion}",
+            "org.slf4j:slf4j-api:${slf4jVersion}",
+            "org.slf4j:slf4j-simple:${slf4jVersion}",
             "org.mongodb:mongodb-driver-sync:4.6.0"
     )
 

--- a/java/src/main/java/app/examples/ElasticsearchIOSimpleRead.java
+++ b/java/src/main/java/app/examples/ElasticsearchIOSimpleRead.java
@@ -47,7 +47,7 @@ public class ElasticsearchIOSimpleRead {
     }
 
     public static void main(String[] args)  {
-        PipelineOptions options = PipelineOptionsFactory.fromArgs().withValidation().create();
+        PipelineOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().create();
         Pipeline pipeline = Pipeline.create(options) ;
         build(pipeline) ;
         pipeline.run().waitUntilFinish() ;

--- a/java/src/main/java/app/examples/ElasticsearchIOSimpleRead.java
+++ b/java/src/main/java/app/examples/ElasticsearchIOSimpleRead.java
@@ -47,7 +47,7 @@ public class ElasticsearchIOSimpleRead {
     }
 
     public static void main(String[] args)  {
-        PipelineOptions options = PipelineOptionsFactory.create();
+        PipelineOptions options = PipelineOptionsFactory.fromArgs().withValidation().create();
         Pipeline pipeline = Pipeline.create(options) ;
         build(pipeline) ;
         pipeline.run().waitUntilFinish() ;

--- a/java/src/main/java/app/examples/ElasticsearchIOSimpleRead.java
+++ b/java/src/main/java/app/examples/ElasticsearchIOSimpleRead.java
@@ -10,12 +10,12 @@ import org.apache.beam.sdk.options.PipelineOptions;
 import org.apache.beam.sdk.options.PipelineOptionsFactory;
 import org.apache.beam.sdk.transforms.MapElements;
 import org.apache.beam.sdk.values.PCollection;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
 
 public class ElasticsearchIOSimpleRead {
 
-    final Logger logger = LogManager.getLogger(
+    final Logger logger = LoggerFactory.getLogger(
             ElasticsearchIOSimpleRead.class) ;
 
     public static void build(Pipeline p) {

--- a/java/src/main/java/app/examples/ElasticsearchIOSimpleWrite.java
+++ b/java/src/main/java/app/examples/ElasticsearchIOSimpleWrite.java
@@ -73,7 +73,7 @@ public class ElasticsearchIOSimpleWrite {
     }
 
     public static void main(String[] args)  {
-        PipelineOptions options = PipelineOptionsFactory.fromArgs().withValidation().create();
+        PipelineOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().create();
         Pipeline pipeline = Pipeline.create(options) ;
         build(pipeline) ;
         pipeline.run().waitUntilFinish() ;

--- a/java/src/main/java/app/examples/ElasticsearchIOSimpleWrite.java
+++ b/java/src/main/java/app/examples/ElasticsearchIOSimpleWrite.java
@@ -69,7 +69,7 @@ public class ElasticsearchIOSimpleWrite {
         ElasticsearchIORepository.write(
                 pCol,
                 con
-);
+        );
     }
 
     public static void main(String[] args)  {
@@ -78,5 +78,4 @@ public class ElasticsearchIOSimpleWrite {
         build(pipeline) ;
         pipeline.run().waitUntilFinish() ;
     }
-
 }

--- a/java/src/main/java/app/examples/ElasticsearchIOSimpleWrite.java
+++ b/java/src/main/java/app/examples/ElasticsearchIOSimpleWrite.java
@@ -15,16 +15,16 @@ import org.apache.beam.sdk.transforms.Create;
 import org.apache.beam.sdk.transforms.MapElements;
 import org.apache.beam.sdk.transforms.SimpleFunction;
 import org.apache.beam.sdk.values.PCollection;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
 
 public class ElasticsearchIOSimpleWrite {
 
-    final Logger logger = LogManager.getLogger(
+    final Logger logger = LoggerFactory.getLogger(
             ElasticsearchIOSimpleWrite.class) ;
 
     private static class StringToJsonFn extends SimpleFunction<String, String> {
-        final Logger logger = LogManager.getLogger(
+        final Logger logger = LoggerFactory.getLogger(
                 StringToJsonFn.class
         );
 

--- a/java/src/main/java/app/examples/ElasticsearchIOSimpleWrite.java
+++ b/java/src/main/java/app/examples/ElasticsearchIOSimpleWrite.java
@@ -73,7 +73,7 @@ public class ElasticsearchIOSimpleWrite {
     }
 
     public static void main(String[] args)  {
-        PipelineOptions options = PipelineOptionsFactory.create();
+        PipelineOptions options = PipelineOptionsFactory.fromArgs().withValidation().create();
         Pipeline pipeline = Pipeline.create(options) ;
         build(pipeline) ;
         pipeline.run().waitUntilFinish() ;

--- a/java/src/main/java/app/examples/MongoIOSimpleQuery.java
+++ b/java/src/main/java/app/examples/MongoIOSimpleQuery.java
@@ -13,8 +13,8 @@ import org.apache.beam.sdk.transforms.MapElements;
 import org.apache.beam.sdk.values.PCollection;
 import org.apache.beam.sdk.values.TypeDescriptor;
 import org.apache.beam.sdk.values.TypeDescriptors;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
 import org.bson.Document;
 
 import java.util.ArrayList;
@@ -23,7 +23,7 @@ import java.util.List;
 import static app.pipelines.connectors.MongoIORepository.ReadWithCustomQuery;
 
 public class MongoIOSimpleQuery {
-    final Logger logger = LogManager.getLogger(
+    final Logger logger = LoggerFactory.getLogger(
             MongoIOSimpleRead.class) ;
 
     /***

--- a/java/src/main/java/app/examples/MongoIOSimpleRead.java
+++ b/java/src/main/java/app/examples/MongoIOSimpleRead.java
@@ -14,15 +14,15 @@ import org.apache.beam.sdk.transforms.MapElements;
 import org.apache.beam.sdk.values.PCollection;
 import org.apache.beam.sdk.values.TypeDescriptor;
 import org.apache.beam.sdk.values.TypeDescriptors;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
 import org.bson.Document;
 
 import static app.pipelines.connectors.MongoIORepository.Read;
 
 public class MongoIOSimpleRead {
 
-    final Logger logger = LogManager.getLogger(
+    final Logger logger = LoggerFactory.getLogger(
             MongoIOSimpleRead.class) ;
 
     /***

--- a/java/src/main/java/app/examples/MongoIOSimpleWrite.java
+++ b/java/src/main/java/app/examples/MongoIOSimpleWrite.java
@@ -12,15 +12,15 @@ import org.apache.beam.sdk.transforms.Create;
 import org.apache.beam.sdk.transforms.MapElements;
 import org.apache.beam.sdk.values.PCollection;
 import org.apache.beam.sdk.values.TypeDescriptor;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
 import org.bson.Document;
 
 import static app.pipelines.connectors.MongoIORepository.Write;
 
 public class MongoIOSimpleWrite {
 
-    final Logger logger = LogManager.getLogger(
+    final Logger logger = LoggerFactory.getLogger(
             MongoIOSimpleWrite.class) ;
 
     /***

--- a/java/src/main/java/app/pipelines/elements/SimpleMongoDocument.java
+++ b/java/src/main/java/app/pipelines/elements/SimpleMongoDocument.java
@@ -2,8 +2,8 @@ package app.pipelines.elements;
 
 import org.apache.beam.sdk.coders.AvroCoder;
 import org.apache.beam.sdk.coders.DefaultCoder;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.Logger;
 import org.bson.Document;
 
 import java.util.HashMap;
@@ -12,7 +12,7 @@ import java.util.Map;
 // https://beam.apache.org/documentation/programming-guide/#annotating-custom-type-default-coder
 @DefaultCoder(AvroCoder.class)
 public class SimpleMongoDocument {
-    private final static Logger logger = LogManager.getLogger(
+    private final static Logger logger = LoggerFactory.getLogger(
             SimpleMongoDocument.class) ;
 
     public String sentence ;

